bigquery/storage/managedwriter/managed_stream.go

```diff
@@ -360,11 +366,29 @@ func (ms *ManagedStream) AppendRows(ctx context.Context, data [][]byte, opts ...
 			ProtoDescriptor: pw.newSchema,
 		}
 	}
-	// proceed to call
-	if err := ms.append(pw); err != nil {
-		// pending write is DOA.
-		pw.markDone(NoStreamOffset, err, ms.fc)
-		return nil, err
+	// Call the underlying append.  The stream has it's own retained context and will surface expiry on
+	// it's own, but we also need to respect any deadline for the provided context.
+	errCh := make(chan error)
+	var appendErr error
+	go func() {
```

Comment:
There is the potential here that `append` will block waiting to add a pending write to the pending writes channel. When that happens the context below could expire returning an error from `AppendRows`. However this goroutine is still running and as soon as the pending channel is freed the `append` will in fact succeed even if an error was returned.

Furthermore this goroutine not being stopped when the caller context expires means that if `append` becomes slow due a full pending write channel, rapid calls to `AppendRows`, even with a short timeouts, could result in a very large number of gorountines left running.

Reply:
Good point.  I was reasoning through backend behaviors, didn't think through this scenario.

Reply:
Spend some time refactoring this, as there were other correctness issues as well primarily around potential races.

Reply:
Also, added a test explicitly to watch for goroutine leak.